File: recursion/prob297.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import (
    "strconv"
    "strings"
    "fmt"
)

type Codec struct {
    
}

func Constructor() Codec {
    return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
    
    if root == nil {
        return ""
    }
    
    ans := strconv.Itoa(root.Val)
    
    if root.Left != nil {
        ans += "[" + this.serialize(root.Left)
        if root.Right != nil {
            ans += " " + this.serialize(root.Right) + "]"
        } else {
            ans += "]"
        }
    } else if root.Right != nil {
        ans += "[ " + this.serialize(root.Right) + "]"
    }
    
    fmt.Println(ans)
    return ans
    
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {    
    
    if len(data) == 0 {
        return nil
    }
    
    if ! strings.Contains(data, "[") {
        num, _ := strconv.Atoi(data)
        root := TreeNode{
            Val: num,
        }
        return &root
    }
    
    l, r := strings.Index(data, "["), strings.LastIndex(data, "]")
    
    num, _ := strconv.Atoi(data[:l])
    root := TreeNode{
        Val: num,
    }
    
    children := helper(data[l+1:r])
    
    left_str, ok := children["left"]
    if ok {
        root.Left = this.deserialize(left_str)
    }
    
    right_str, ok := children["right"]
    if ok {
        root.Right = this.deserialize(right_str)
    }
    
    return &root
    
}

func helper(s string) map[string]string {
    
    ans := map[string]string{}
    
    if s[0] == ' ' {
        ans["right"] = s[1:]
        return ans
    }
    
    stack := 0
    n := len(s)
    for i := 0; i < n; i++ {
        if s[i] == '[' {
            stack++
        } else if s[i] == ']' {
            stack--
        } else if s[i] == ' ' {
            if stack == 0 {
                ans["left"] = s[:i]
                ans["right"] = s[i+1:]
                return ans
            }
        }
    }
    
    ans["left"] = s
    
    return ans
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */

File: recursion/prob297.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        ans = str(root.val)
        if root.left or root.right:
            if root.left and root.right:
                children = f"[{self.serialize(root.left)} {self.serialize(root.right)}]"
            elif root.left:
                children = f"[{self.serialize(root.left)}]"
            else:
                children = f"[ {self.serialize(root.right)}]"
            ans += children
            
        #print(ans)
        return ans
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        if "[" not in data:
            return TreeNode(int(data))
        
        l = data.index("[")
        r = data.rindex("]")
        root = TreeNode(int(data[:l]))
        
        children_str = data[l+1:r]
        children = self.helper(children_str)
        if "left" in children:
            root.left = self.deserialize(children["left"])
        if "right" in children:
            root.right = self.deserialize(children["right"])
        
        return root
        
        
        
    def helper(self, s):
        "split the children string"
        d = {}
        if s[0] == " ":  # only right child
            d["right"] = s[1:]
            return d
        
        # find the splitting " " char
        n = len(s)
        stack = 0
        for i in range(n):
            if s[i] == "[":
                stack += 1
            elif s[i] == "]":
                stack -= 1
            elif s[i] == " ":
                if stack == 0:
                    d["left"] = s[:i]
                    d["right"] = s[i+1:]
                    return d
                
        d["left"] = s  # only left child
        return d
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
